File: c/Dota2 Senate.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    int* data;
    int front;
    int rear;
    int capacity;
} Queue;

void initQueue(Queue* q, int capacity) {
    q->data = (int*)malloc(capacity * sizeof(int));
    q->front = 0;
    q->rear = 0;
    q->capacity = capacity;
}

int isEmpty(Queue* q) {
    return q->front == q->rear;
}

void enqueue(Queue* q, int value) {
    q->data[q->rear++] = value;
}

int dequeue(Queue* q) {
    return q->data[q->front++];
}

void freeQueue(Queue* q) {
    free(q->data);
}

char* predictPartyVictory(char* senate) {
    int n = strlen(senate);
    Queue radiant, dire;
    initQueue(&radiant, n);
    initQueue(&dire, n);

    for (int i = 0; i < n; ++i) {
        if (senate[i] == 'R') {
            enqueue(&radiant, i);
        } else {
            enqueue(&dire, i);
        }
    }

    while (!isEmpty(&radiant) && !isEmpty(&dire)) {
        int r = dequeue(&radiant);
        int d = dequeue(&dire);

        if (r < d) {
            enqueue(&radiant, r + n);
        } else {
            enqueue(&dire, d + n);
        }
    }

    char* result = (char*)malloc(8 * sizeof(char));
    if (!isEmpty(&radiant)) {
        strcpy(result, "Radiant");
    } else {
        strcpy(result, "Dire");
    }

    freeQueue(&radiant);
    freeQueue(&dire);

    return result;
}
//  test case
int main() {
    char senate[] = "RDDRR";
    printf("result: %s\n", predictPartyVictory(senate));
    return 0;
}

File: python/Dota2 Senate.py
from collections import deque

class Solution(object):
    def predictPartyVictory(self, senate):
        radiant = deque()
        dire = deque()

        for i, s in enumerate(senate):
            if s == 'R':
                radiant.append(i)
            else:
                dire.append(i)

        n = len(senate)

        while radiant and dire:
            r = radiant.popleft()
            d = dire.popleft()
            
            if r < d:
                radiant.append(r + n)
            else:
                dire.append(d + n)
        
        return "Radiant" if radiant else "Dire"

# Test cases
if __name__ == "__main__":
    solution = Solution()
    test_senate_1 = "RD"
    test_senate_2 = "RDD"
    
    print(solution.predictPartyVictory(test_senate_1))  
    print(solution.predictPartyVictory(test_senate_2))  

